feat: реализован header (без бургер меню) и notfound. в процессе: рег…

OlgaSolo21 opened the pull request:
это результат моей работы, с учетом отсутствия сна в ночь с пятницы на субботу))

OlgaSolo21 committed:
feat: реализован header (без бургер меню) и notfound. в процессе: регистрация

OlgaSolo21 committed:
feat: регистрация/логин с токенами

OlgaSolo21 committed:
feat: currentUser и profile

OlgaSolo21 committed:
feat: профиль изменен - доработать ui и useForm

OlgaSolo21 committed:
feat: профиль доработка

OlgaSolo21 committed:
feat: роуты на юзера
progress: роут на фильмы и форма поиска, без фильтра

OlgaSolo21 committed:
разбор роута movies, ощибка в фильтрации данных в utils

OlgaSolo21 committed:
настройка фильтрации фильмов - разобрать доп на русском не фильтрует

OlgaSolo21 committed:
геаризация поика и записи в сторадж, проверить монтирование

OlgaSolo21 committed:
все тз без сна: реализован весь функционал

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [ ] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске на странице фильмов текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными.
    **Комментарий: пользователь не видит точного результата своего последнего поиска на странице фильмов при повторном переходе на эту страницу с другой страницы и после перезагрузки сайта - видит отображение всех фильмов, соответствующих строке поиска, даже, если перед уходом или перезагрузкой были найдены короткометражки.**
    
    **Комментарий (пояснение, как должен быть реализован функционал повторного показа результата последнего поиска на странице фильмов): для того, чтобы выполнить вышележащий пункт чек-листа, при каждом поиске на странице фильмов результаты поиска нужно сохранять в localstorage. В сохраняемые данные должны входить строка поиска и статус чекбокса «Короткометражки» - логическое значение, показыающее в активном ("короткометражки") или дефолтном положении находится чекбокс. При появлении пользователя на странице фильмов вновь после того, как он перезагрузит сайт, или перейдёт на страницу фильмов с другой страницы, сохранённую в локальном хранилище информацию нужно будет использовать для вывода на экран результата поиска, который был перед перезагрузкой или уходом на другую страницу (с расстановкой актуального состояния лайков). При логауте пользователя необходимо эту информацию в хранилище  и стейт-переменных из useState уничтожать.** 
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
	**Комментарий: сообщения валидации полей ввода на формах аутентификации должны отображаться или исчезать при каждом изменении значения поля (при событии onChange): при вводе пользователем символов в поля, заполнении поля по свойству autocomplete или с помощью копирования значений. Сейчас эти сообщения под полями ввода отображаются только при потере полем фокуса, что не удобно для пользователя, так как он может видеть сообщение об ошибке, когда поле уже валидно, и не видеть сообщения в тот момент, когда поле не валидно (см. снимки). Кнопка сабмита должна становиться активной, как только значения всех полей ввода станут валидными, без каких-либо дополнитеьых действий пользователя, и сразу же становиться или оставаться неактивной, если значение хотя бы в каком-то из полей ввода не валидно. При событии onBlur или onFocus сделать валидацию правильно гораздо труднее.**
  - Страницы «Регистрация» и «Авторизация»:  
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    **Комментарий: На странице профиля после попытки сохранить адрес почты, который уже был зарегистрирован, происходит редирект на главную страницу (см. снимок). Но, если почта уже была зарегистрирована,  пользователь так и должен оставаться на странице профиля, иметь возможность исправить данные и попробовать сохранить изменённые данные снова.**
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.
	**Комментарий (Нужно протестировать функционал, введя новые данные профиля, например адрес почты, а затем сразу (до отсылки запроса на сервер) исправив их на текущие данные): При редактировании пользователь может ввести текущие данные профиля и сохранить их, а нужно сделать так, чтобы можно было отправить запрос на сохранение, только если  данные изменены по сравнению с текущими данными пользователя, которые должны содержаться в текущий момент в контексте.**
	
	**Напоминание: после каждого успешного сохранения данных профиля на сервере, нужно эти данные вносить в контекст юзера.**
	
	**Напоминание: Чтобы данные профиля считались изменёнными, пользователю должно быть достаточно изменения хотя бы одного данного. Может быть изменён один любой инпут из двух, и отправлен запрос на сервер.**  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
	  **Комментарий: При выводе результата поиска при передвижении чекбокса, без нажатия кнопки "Поиск" на форме поиска, должен работать тот же алгоритм, который работает при сабмите формы поиска, то есть работает при нажатии кнопки "Поиск". Таким образом, при передвижении чекбокса для получения результата поиска должен производиться новый фильтр всего массива фильмов по заданной в текущий момент строке поиска в форме поиска и положению самого чекбокса. Сейчас при передвижении чекбокса для вывода результата берётся строка поиска, которая была при предыдущем поиске, а не новая строка, которую ввёл пользователь перед передвижением чекбокса (см. снимок). Сделайте в соответствии с этим замечанием и алгоритм поиска на странице «Сохранённые фильмы».**
	  
	**Комментарий: данные о результате поиска со страницы фильмов должны сохраняться в localstorage при каждом поиске, независимо от того каким способом произведён поиск, то есть и при выводе результата поиска от нажатия кнопки "Поиск", и при выводе результата поиска вследствие передвижения чекбокса.**  
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  **Комментарий: Вывод карточек на экран не соответствует постановке задачи, данной в описании диплома, хотя бы потому, что на экран всегда должны выводиться одинаковые по количеству карточек ряды карточек ( возможно, кроме последнего ряда), а сейчас это не так (см. снимок при ширине экрана 1250px). На снимке приведён пример неправильного вывода карточек. Нужно протестировать и отладить работу функционала в соответствии с приведёнными ниже требованиями и подсказками (в пунктах 1 и 2) при различных действиях пользователя и разной ширине экрана.**

**Ориентироваться нужно на описание функционала рендера карточек, которое дано в задании по диплому:**
   Обратите внимание, что количество карточек, которые отображаются на странице, зависит от ширины экрана устройства.
    • Ширина 1280px — 4 ряда карточек. Кнопка «Ещё» загружает дополнительный ряд карточек.
    • Ширина 768px — 4 ряда карточек. Кнопка «Ещё» загружает ряд карточек.
    • Ширина от 320px до 480px — 5 карточек по 1 в ряд. Кнопка «Ещё» загружает по 2 карточки.
В ряду отображайте столько карточек, сколько отображается в вашем макете на соответствующей ширине экрана.
  
  **Комментарий (пояснение, как должен быть реализован функционал вывода карточек, чтобы удовлетворить требованиям, приведённым в цитате выше):**  
  **1. При изменении пользователем ширины экрана результат поиска нужно выводить с самого начала с применением дебаунса. То есть, когда пользователь остановится на какой-то ширине экрана, результат нужно выводить снова, начиная с первых 4-х или 5-ти рядов, после которых вывести кнопку "Ещё". Результат выводить в соответствии с правилами для текущей ширины экрана.**  

  **2. При осуществлении нового поиска на той же ширине экрана, при которой был предыдущий поиск, вывод карточек нужно начинать с самого начала схемы вывода карточек, то есть с первых 4-х (или 5-ти) рядов, после которых нужно вывести кнопку "Ещё".**
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
    **Комментарий: При клике на активную кнопку лайка на странице фильмов, кнопка  должна становиться неактивной, а сама карточка должна удаляться из сохранённых и исчезать с экрана на странице «Сохранённые фильмы». Сейчас происходит наоборот: карточка не удаляется из избранных, а в избранные добавляется ещё экземпляр этой карточки.**
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [ ] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  **Комментарий: При попытке авторизованного пользователя войти на любую приватную страницу по её URL-адресу в адресной строке, происходит вход пользователя на страницу «Фильмы», а не на эти страницы, хотя редирект не по набранному адресу должен происходить только при попытке авторизованного пользователя выйти на страницы авторизации и регистрации.**
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [ ] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  **Комментарий: авторизованные пользователи не могут войти на страницу «404».**
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [ ] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
- [ ] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
**Комментарий: Все исходные числовые данные проекта - длительность короткометражек, количество фильмов, отображаемое при нажатии кнопки "Еще", другие постоянные числовые данные, значения которым задаются до начала выполнения приложения, нужно выносить в отдельный файл констант, чтобы эти исходные данные было легко менять. В коде принято использовать только имена этих констант, чтобы не было "магических" значений.**
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 57.
**P.S.**
**Все исправления кода должны быть видны в коммитах в репозитории. На исправление каждой ошибки должен быть отдельный коммит. Сделайте, пожалуйста, у каждого коммита понятное название, говорящее, что коммит исправляет.**

OlgaSolo21 committed:
bugfix:
1.  константы именуются заглавными буквами и вынесены в отдельный конфиг-файл.
2. Вывод карточек на экран при различных действиях пользователя и разной ширине экрана

OlgaSolo21 committed:
bugfix:
1.Сохранение в localStoradge результаты поиска и статус чекбокса
1.1 данные при смене страницы и возвращении на "Фильмы" берутся из хранилища, втом числе лайки

OlgaSolo21 committed:
bugfix:
1.Валидация полей ввода на формах аутентификации (показать/скрыть ошибку)
1.1  Кнопка сабмита меняет состояние в зависимости от валидации полей

OlgaSolo21 committed:
bugfix:Изменение Profile
1. при 409 ошибке не происходит выход из аккаунта
2. после каждого успешного сохранения данных они вносятся в контекст юзера.
3. тест функционала кнопки "сохранить" - состояние

OlgaSolo21 committed:
чистка кода

OlgaSolo21 committed:
чистка кода

OlgaSolo21 committed:
bugfix: чекбокс
1. при передвижении чекбокса, его статус устанавливается в зависимости от нажатия

OlgaSolo21 committed:
bugfix: сохранение/удаление фильма: При клике на активную кнопку лайка на странице фильмов, кнопка должна становиться неактивной, а сама карточка должна удаляться из сохранённых и исчезать с экрана на странице «Сохранённые фильмы»

OlgaSolo21 committed:
bugfix:Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации

OlgaSolo21 commented:
- [ ] Комментарий: авторизованные пользователи не могут войти на страницу «404».

OlgaSolo21 commented:
- [ ] Комментарий: авторизованные пользователи не могут войти на страницу «404».
_""не совсем поняла Вас, если в адресной строке ввести не верный адрес авторизованный пользователь также попадает на страницу 404
![notFoundRes](https://github.com/OlgaSolo21/movies-explorer-frontend/assets/132297852/056dae09-ac4f-418b-bd0d-b6653ebda720)
![notFoundReq](https://github.com/OlgaSolo21/movies-explorer-frontend/assets/132297852/6d4b7000-49cc-45ac-a91d-dfe467f255f3)
_

OlgaSolo21 commented:
- [ ] Комментарий: данные о результате поиска со страницы фильмов должны сохраняться в localstorage при каждом поиске, независимо от того каким способом произведён поиск, то есть и при выводе результата поиска от нажатия кнопки "Поиск", и при выводе результата поиска вследствие передвижения чекбокса. 
- [ ] 

> Тут я не знаю, что и думать:) перепробовала кучу вариантов - не могу сделать так чтобы чекбокс работал как сабмит поиска. Если подскажете в какую сторону думать, поправлю все быстро, до дедлайна. 
Спасибо!

irinamozes commented:
Здравствуйте!

Нужно рассмотреть событие "change" чекбокса и добавить на чекбокс обработчик этого события. Код этого обработчика должен быть примерно такой же как и у обработчика события нажатия кнопки "Поиск" на форме поиска. При этом при выводе результата поиска в этом обработчике нужно учитывать, что введено в форму поиска в текущий момент и в каком положении находится сам чекбокс. Простой и хороший пример для добавления обработчика на чекбокс есть здесь: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox.

OlgaSolo21 commented:
> @OlgaSolo21 Здравствуйте!
> 
> Нужно рассмотреть событие "change" чекбокса и добавить на чекбокс обработчик этого события. Код этого обработчика должен быть примерно такой же как и у обработчика события нажатия кнопки "Поиск" на форме поиска. При этом при выводе результата поиска в этом обработчике нужно учитывать, что введено в форму поиска в текущий момент и в каком положении находится сам чекбокс. Простой и хороший пример для добавления обработчика на чекбокс есть здесь: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox.

> поняла Вас, спасибо! я тогда втечение 3,5 часов переделаю и залью изменения (сейчас не у компа)

OlgaSolo21 committed:
bugfix:реализация сабмита на чекбокс с учетом состояния

OlgaSolo21 commented:
@irinamozes 
чекбокс реализовала, сейчас направлю новую версию Вам. спасибо!

irinamozes commented:
Добрый вечер, Ольга.
Исправление работы чекбокса пока не получилось.
Прочитайте, пожалуйста, внимательно все комментарии (и в которых указывается ошибка, и в которых даются пояснения, как ошибку исправлять) из резюме, расположенном ниже.
Работу чекбокса нужно исправить после исправления сохранения данных поиска в localStorage (первый комментарий и пояснения к нему). 

# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [ ] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске на странице фильмов текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными.
    **Комментарий: пользователь не видит точного результата своего последнего поиска на странице фильмов при повторном переходе на эту страницу с другой страницы и после перезагрузки сайта. Данное shortsCheckbox ⚠️не меняет своего значения ни при выводе результата вследствие нажатия кнопки "Поиск", ни при выводе результата вследствие передвижения чекбокса в другое положение. Ключ shortsCheckbox установился в "true" при первом поиске короткометражек, и больше его значение не менялось (см. снимок).**
  
![Снимок экрана в 2024-03-17 23-34-09](https://github.com/OlgaSolo21/movies-explorer-frontend/assets/19558891/ce474dc9-6991-4b65-b16b-85ef7fa88a76)
  
    **Как указывалось при прошлой проверке:**
    данные о результате поиска со страницы фильмов должны сохраняться в localstorage при каждом поиске, независимо от того каким способом произведён поиск, то есть и при выводе результата поиска от нажатия кнопки "Поиск", и при выводе результата поиска вследствие передвижения чекбокса.
    **При повторном показе последнего поиска, нужно выводить на экран результат поиска в ⚠️точном соответствии с данными, которые содержатся в localstorage.**
    
    **Комментарий (⚠️пояснение, как должен быть реализован функционал повторного показа результата последнего поиска на странице фильмов): для того, чтобы выполнить вышележащий пункт чек-листа, при каждом поиске на странице фильмов результаты поиска нужно сохранять в localstorage. В сохраняемые данные должны входить строка поиска и статус чекбокса «Короткометражки» - логическое значение, показыающее в активном ("короткометражки") или дефолтном положении находится чекбокс. При появлении пользователя на странице фильмов вновь после того, как он перезагрузит сайт, или перейдёт на страницу фильмов с другой страницы, сохранённую в локальном хранилище информацию нужно будет использовать для вывода на экран результата поиска, который был перед перезагрузкой или уходом на другую страницу (с расстановкой актуального состояния лайков). При логауте пользователя необходимо эту информацию в хранилище  и стейт-переменных из useState уничтожать.** 
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
  - Страницы «Регистрация» и «Авторизация»:  
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
	  **Комментарий (исправления этого пункта не получилось. Нужно проверить, какой обработчик Вы устанавливаете на атрибут onChange чекбокса. Именно об этом обработчике я Вам и писала в ответе на Ваш вопрос. Сейчас в этом атрибуте стоит обработчик toggleCheckBox. Но, выводит ли он результат поиска?): При выводе результата поиска при передвижении чекбокса, без нажатия кнопки "Поиск" на форме поиска, должен работать тот же алгоритм, который работает при сабмите формы поиска, то есть работает при нажатии кнопки "Поиск". Таким образом, при передвижении чекбокса для получения результата поиска и вывода его на экран должен производиться новый фильтр всего массива фильмов по ⚠️заданной в текущий момент строке поиска в форме поиска и ⚠️положению (или статусу) самого чекбокса.**
	  
	**Комментарий (замечание не исправлено): данные о результате поиска со страницы фильмов должны сохраняться в localstorage при каждом поиске, независимо от того каким способом произведён поиск, то есть и при выводе результата поиска от нажатия кнопки "Поиск", и при выводе результата поиска вследствие передвижения чекбокса.**  
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
    **Комментарий: Вывод карточек на экран не вполне соответствует постановке задачи, данной в описании диплома. Нужно протестировать и отладить работу функционала в соответствии с приведёнными ниже требованиями и подсказками (в пунктах 1 и 2) при различных действиях пользователя и разной ширине экрана.**

**Ориентироваться нужно на описание функционала рендера карточек, которое дано в задании по диплому:**
   Обратите внимание, что количество карточек, которые отображаются на странице, зависит от ширины экрана устройства.
    • Ширина 1280px — 4 ряда карточек. Кнопка «Ещё» загружает дополнительный ряд карточек.
    • Ширина 768px — 4 ряда карточек. Кнопка «Ещё» загружает ряд карточек.
    • Ширина от 320px до 480px — 5 карточек по 1 в ряд. Кнопка «Ещё» загружает по 2 карточки.
В ряду отображайте столько карточек, сколько отображается в вашем макете на соответствующей ширине экрана.
  
  **Комментарий (пояснение, как должен быть реализован функционал вывода карточек, чтобы удовлетворить требованиям, приведённым в цитате выше):**  
  **1(этот пункт выполняется, но нужно ещё его протестировать при различных действиях пользователя после исправления второго пункта). При изменении пользователем ширины экрана результат поиска нужно выводить с самого начала с применением дебаунса. То есть, когда пользователь остановится на какой-то ширине экрана, результат нужно выводить снова, начиная с первых 4-х или 5-ти рядов, после которых вывести кнопку "Ещё". Результат выводить в соответствии с правилами для текущей ширины экрана.**
  **2 (этот пункт не выполняется). При осуществлении нового поиска на экран сейчас выводится до появления кнопки «Ещё» сразу столько рядов карточек, сколько было выведено при предыдущем поиске. Но, при новом поиске вывод карточек нужно начинать с самого начала схемы вывода карточек, то есть с первых 4-х (или 5-ти) рядов, после которых нужно вывести кнопку "Ещё".**
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    **Комментарий: сегодня в форму поиска на странице «Сохранённые фильмы» ничего нельзя было ввести. Поиск на странице «Сохранённые фильмы» не работал.**
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [ ] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  **Комментарий: По URL-адресу авторизованный пользователь может зайти и на страницы авторизации и регистрации, что не допустимо, так как открывается возможность авторизоваться ещё раз, не выйдя из предыдущего аккаунта.**
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.  
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [ ] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
- [ ] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
**Комментарий: длительность короткометражек - число 40, нужно занести в константу с понятным названием и поместить в файл констант, чтобы в коде не было "магических" значений. Смысл литеральных числовых значений  (таких как 40, или любых других чисел), встречающихся в коде, не возможно понять, если их не занести в константы с "говорящим названием" и использовать вместо них эти константы в коде. Поэтому числа и называются "магическими" значениями.**
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 59.
**P.S.**
**Все исправления кода должны быть видны в коммитах в репозитории. На исправление каждой ошибки должен быть отдельный коммит. Сделайте, пожалуйста, у каждого коммита понятное название, говорящее, что коммит исправляет.**

OlgaSolo21 committed:
bugfix:
1.Фикс значение короткометражек вынесена в константу
2. Поля формы заблокированы во время отправки запросов
3. По URL-адресу авторизованный пользователь  НЕ может зайти страницы авторизации и регистрации

OlgaSolo21 committed:
bugfix:
1.При поиске на странице фильмов текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище
2. Вывод карточек на экран (в пунктах 1 и 2)

OlgaSolo21 committed:
bugfix:
1.При поиске на странице фильмов текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище
2. Вывод карточек на экран (в пунктах 1 и 2)

OlgaSolo21 commented:
Доброе утро, Ирина! 
я поправила абсолютно все комментарии и замечания, кроме сабмита чекбокса :(
потратив на эту функцию половину прошлой ночи, вчерашний день и сегодняшнюю ночь, я к сожалению так и не пришла к решению. 
Также уточняла вчера у ребят из своей когорты - почти ни у кого из тех, с кем общалась не требовали выполнения данного пункта (именно поиск по чекбоксу)
К сожалению, время дедлайна 12:00 18/03 как мне сказала куратор, поэтому я очень попрошу Вас дать мне реализовать этот функционал за пределами дедлайна, а на данный момент принять диплом в таком формате, ведь все остальное по требованиям ТЗ приложения работает (я посчитала, там как раз на 80 баллов хватает, если без этого пункта :)) 
Надеюсь на Ваше понимание.
Спасибо!

irinamozes commented:
@OlgaSolo21 
> Также уточняла вчера у ребят из своей когорты - почти ни у кого из тех, с кем общалась не требовали выполнения данного пункта (именно поиск по чекбоксу)

Возможно, такого прямого требования не было года 1.5 назад. Но поменялось ТЗ, появилось и требование, которое логично, иначе пользователь получает не то, что ожидал. Сейчас дипломы по функционалу проверяют 3 ревьюера, 2 из них точно выполнение этого пункта требуют, поскольку это обязательное требование. Третий тоже должен, но не спрашивала.

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске на странице фильмов текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными. 
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
  - Страницы «Регистрация» и «Авторизация»:  
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
	  **Комментарий (исправления этого пункта не получилось). При выводе результата поиска при передвижении чекбокса, без нажатия кнопки "Поиск" на форме поиска, должен работать тот же алгоритм, который работает при сабмите формы поиска, то есть работает при нажатии кнопки "Поиск".**	   
	  
	**Комментарий (замечание не исправлено): данные о результате поиска со страницы фильмов должны сохраняться в localstorage при каждом поиске, независимо от того каким способом произведён поиск, то есть и при выводе результата поиска от нажатия кнопки "Поиск", и при выводе результата поиска вследствие передвижения чекбокса.**  
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.  
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [x] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
- [x] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 94.
6 баллов снято за замечание из раздела "Функциональность проекта из брифа реализована полностью".
## У Вас очень хороший дипломный проект!

Чтобы красиво оформить файл README.md, где рассказать о собственной работе над проектом и собственных достижениях, можно воспользоваться инструментом Markdown. Описание языка Markdown есть в Ваших лекциях по этой ссылке:
https://praktikum.yandex.ru/learn/web/courses/dbf98e55-0f76-444b-850c-4538708ad571/sprints/1425/topics/b4072eed-2089-45c0-9382-98ea71202341/lessons/939aeab7-0508-49e8-bcf2-199d4dbf74f7/.

Ссылки на интересные материалы и подкасты:
[habr](https://habr.com/ru/)
https://soundcloud.com/everyonecan
https://soundcloud.com/podlodka
https://soundcloud.com/devschacht
https://soundcloud.com/begebot
https://soundcloud.com/web-standards
https://www.youtube.com/channel/UCZeU17nbVfzczAkJVTay9vw

Интересная статья о замыкании и лексическом окружении функций:
https://developer.mozilla.org/ru/docs/Web/JavaScript/Closures.

О пользовательских хуках:
https://ru.reactjs.org/docs/hooks-custom.html

https://reactdev.ru/
Справочник React

Какие бывают стили в названиях файлов js и зачем они нужны, интересно и кратко написано здесь:
https://medium.com/@alivander/camel-pascal-snake-case-и-другие-стили-написания-288ec62ca0d0

Ресурс по библиотеке Redux, которая очень широко используется в реальных проектах на React:
https://rajdee.gitbooks.io/redux-in-russian/content/

Основы TypeScript
https://code-basics.com/ru/languages/typescript

Создание приложений JavaScript с помощью TypeScript
https://docs.microsoft.com/ru-ru/learn/paths/build-javascript-applications-typescript/

Борис Черный
Профессиональный TypeScript. Разработка масштабируемых JavaScript-приложений

**Желаю дальнейших творческих успехов, экспертных знаний и удачи!**
